### Importing Data set 

In [1]:
import pandas as pd
wine_data = pd.read_csv(r'C:\temp\winequality-red.csv', 
                        names=['Fixed_Acidity', 
                               'Volatile_Acidity', 
                               'Citric_Acid', 
                               'Residual_Sugar', 
                               'Chlorides', 
                               'Free_Sulfur_Dioxide', 
                               'Total_Sulfur_Dioxide', 
                               'Density', 
                               'pH', 
                               'Sulphates', 
                               'Alcohol', 
                               'Quality'
                              ],
                        skiprows=1,
                        sep=r'\s*;\s*', engine='python')

In [2]:
wine_data.head()

Fixed_Acidity  Volatile_Acidity  Citric_Acid  Residual_Sugar  Chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   Free_Sulfur_Dioxide  Total_Sulfur_Dioxide  Density    pH  Sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   Alcohol  Quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

### Preprocessing  Checking for NUlls ETC...  I have done this in another notebook - 

### I am going to assume I have all the data clean 

In [5]:
X = wine_data.drop('Quality', 1)
y = wine_data['Quality']

### splitting the data set 

In [6]:
# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

### PCA performs best with a normalized feature set. We will perform standard scalar normalization to normalize our feature set. To do this, execute the following code:

In [7]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [25]:
print(len(X_train))
print(len(X_test))


1279
320


 ### we create a PCA object named pca. We did not specify the number of components in the constructor. Hence, all 11  of the features in the feature set will be returned for both the training and test sets.

In [8]:
from sklearn.decomposition import PCA

pca = PCA()
X_train = pca.fit_transform(X_train)
X_test = pca.transform(X_test)

In [9]:
explained_variance = pca.explained_variance_ratio_

In [10]:
explained_variance

array([0.28263119, 0.17942214, 0.1358536 , 0.10862763, 0.08692731,
       0.06008452, 0.0533286 , 0.03926827, 0.03166857, 0.01667793,
       0.00551024])

It can be seen that first principal component is responsible for 28.26% variance. 
Similarly, the second principal component causes 17.9% variance in the dataset....etc... Collectively we can say that (28.26+17.9+13.58+10.86+8.7) 79.3% percent of the classification information contained in the feature set is captured by the first 5 principal components.

Let's first try to use 1 principal component to train our algorithm. To do so, execute the following code

In [34]:
from sklearn.decomposition import PCA

pca = PCA(n_components=1)
X_train = pca.fit_transform(X_train)
X_test = pca.transform(X_test)

### Training and Making Predictions
### we'll use random forest classification to make a prediction 

In [35]:
from sklearn.ensemble import RandomForestClassifier

classifier = RandomForestClassifier(n_estimators=100,max_depth=4, random_state=0)
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

In [36]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

cm = confusion_matrix(y_test, y_pred)
print(cm)
print('Accuracy',accuracy_score(y_test, y_pred))

[[  0   0   1   1   0   0]
 [  0   0   8   3   0   0]
 [  0   0 109  26   0   0]
 [  0   0  99  43   0   0]
 [  0   0  20   7   0   0]
 [  0   0   0   3   0   0]]
Accuracy 0.475


The output that with only one feature, the random forest algorithm is able to correctly predict 143 out of 319 instances, resulting in 45.93% accuracy

#### Now we are going to try with 2,3,4,5 PCA 

In [40]:
from sklearn.decomposition import PCA

pca= PCA(n_components=2)
X_train = pca.fit_transform(X_train)
X_test = pca.transform(X_test)


ValueError: n_components=2 must be between 0 and min(n_samples, n_features)=1 with svd_solver='full'